doc/specs/drafts/#5916 - Triggers/#5916 - Triggers.md

```diff
@@ -0,0 +1,601 @@
+
+
+![image](https://user-images.githubusercontent.com/18356694/185231958-46c73bcc-4ef4-4dfe-a631-593649a64b75.png)
+
+_fig 2: A trigger that adds a red error mark whenever an app writes "error" to the Terminal_
+
+
+![gh-5916-prototype-001](https://user-images.githubusercontent.com/18356694/185247128-b3b5c88a-e46c-4a9c-8a03-fd45b00ebc70.gif)
+
+_fig 3: A sendInput trigger which sends some of the matched text_
+
+
+## Solution Design
+
+To support "triggers", we'll add a new object to the profile settings called
+`triggers`. This is an array of objects that describe
+
+1. What string to match on for running the trigger
+2. What to do when that match is found.
+
+Let's start with an example blob of JSON:
+
+```jsonc
+"triggers":[
+    {
+        "match": "[Ee][Rr][Rr][Oo][Rr]",
+        "action": "addMark",
+        "color": "#ff0000"
+    },
+    {
+        "match": "Terminate batch job \\(Y/N\\)",
+        "action": "sendInput",
+        "input": "y\r\n"
+    },
+    {
+        "match": {
+            "anchor": "bottom",
+            "offset": 4,
+            "length": 5,
+            "pattern": "    git push --set-upstream origin ([^\\w]*)",
+            "runOn": "mark"
+        },
+        "action": "sendInput",
+        "input": "git push --set-upstream origin ${match[1]}\r"
+    },
+    {
+        "match": "'(.*)' is not recognized as an internal or external command,",
+        "action": "checkWinGet",
+        "commandline": "${match[1]}"
+    },
+    {
+        "match": "This will open a new Terminal tab",
+        "action": "newTab"
+    },
+]
+```
+
+`triggers` is an array of `Trigger` objects, which are a combination of an
+`ActionAndArgs`, a `Control.Matcher`, and a `Control.TriggerAction`.
+
+> **Note**
+> Interface and other types here are shown in C#, since that's fairly close to
+> MIDL3, which would be used for cross-component interfaces.
+
+```c#
+runtimeclass Terminal.Settings.Model.Trigger
+{
+    String Id;
+
+    Control.Matcher Matcher;
+    Control.TriggerAction ControlAction;
+
+    Terminal.Settings.Model.ActionAndArgs ParseMatches(String[] matches);
+
+    private String _actionJson;
+}
+```
+
+### Trigger matchers
+
+The `match` property is parsed as a `Control.Matcher`. This includes the
+following properties, which are heavily inspired by the matchers in VsCode.
+
+```jsonc
+"match": {
+    "anchor": "bottom",
+    "offset": 0, // starting how many rows from the bottom
+    "length": 5, // how many lines to match against
+    "pattern": "    git push --set-upstream origin ([^\\w]*)",
+    "runOn": "everything|newline|mark",
+    "buffer": "main|alt|any"
+}
+```
+
+* `"anchor"` (_default: `"bottom"`_) : Which part of the viewport to run the match against.
+  - VsCode has this, but I don't think we really do. I think we can just always say bottom.
+* `"offset"` (_default: `0`_) : Run the match starting how many rows from the anchor
+* `"length"` (_default: `1`_) : How many rows to include in the match

```

Comment:
So does this mean it would match 1 line below the viewport? I don't really get this...